File: device/cuda/causal/native/intervention.c
#include "intervention.h"
#include "causal_dispatch.h"
#include "../internal/bridge/core_private.h"

int cuda_dispatch_do_intervene(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef adjacencyRef,
    CUDABufferRef intervenedRef,
    CUDABufferRef outRef,
    uint32_t nodeCount,
    uint32_t intervenedCount,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (adjacencyRef == NULL || intervenedRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* adjacencyPtr = cuda_buffer_device_ptr(adjacencyRef);
    void* intervenedPtr = cuda_buffer_device_ptr(intervenedRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&adjacencyPtr, &intervenedPtr, &outPtr, &nodeCount, &intervenedCount};

    return cuda_causal_named_launch(
        contextRef,
        elementDType,
        "do_intervene",
        nodeCount * nodeCount,
        args,
        sizeof(args),
        completionToken,
        status
    );
}

int cuda_dispatch_cate(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef treatedRef,
    CUDABufferRef controlRef,
    CUDABufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (treatedRef == NULL || controlRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* treatedPtr = cuda_buffer_device_ptr(treatedRef);
    void* controlPtr = cuda_buffer_device_ptr(controlRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&treatedPtr, &controlPtr, &outPtr, &count};

    return cuda_causal_named_launch(
        contextRef,
        elementDType,
        "cate",
        count,
        args,
        sizeof(args),
        completionToken,
        status
    );
}

int cuda_dispatch_counterfactual(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef observedYRef,
    CUDABufferRef observedXRef,
    CUDABufferRef counterfactualXRef,
    CUDABufferRef slopeRef,
    CUDABufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (observedYRef == NULL || observedXRef == NULL || counterfactualXRef == NULL ||
        slopeRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* observedYPtr = cuda_buffer_device_ptr(observedYRef);
    void* observedXPtr = cuda_buffer_device_ptr(observedXRef);
    void* counterfactualXPtr = cuda_buffer_device_ptr(counterfactualXRef);
    void* slopePtr = cuda_buffer_device_ptr(slopeRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {
        &observedYPtr,
        &observedXPtr,
        &counterfactualXPtr,
        &slopePtr,
        &outPtr,
        &count,
    };

    return cuda_causal_named_launch(
        contextRef,
        elementDType,
        "counterfactual",
        count,
        args,
        sizeof(args),
        completionToken,
        status
    );
}

int cuda_dispatch_iv_estimate(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef instrumentRef,
    CUDABufferRef treatmentRef,
    CUDABufferRef outcomeRef,
    CUDABufferRef scratchRef,
    CUDABufferRef outRef,
    uint32_t count,
    uint32_t partialCount,
    uint64_t completionToken,
    CUDAStatus* status
) {
    if (instrumentRef == NULL || treatmentRef == NULL || outcomeRef == NULL ||
        scratchRef == NULL || outRef == NULL) {
        cuda_causal_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    void* instrumentPtr = cuda_buffer_device_ptr(instrumentRef);
    void* treatmentPtr = cuda_buffer_device_ptr(treatmentRef);
    void* outcomePtr = cuda_buffer_device_ptr(outcomeRef);
    void* scratchPtr = cuda_buffer_device_ptr(scratchRef);
    void* partialArgs[] = {&instrumentPtr, &treatmentPtr, &outcomePtr, &scratchPtr, &count};

    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* finalizeArgs[] = {&scratchPtr, &outPtr, &count, &partialCount};

    return cuda_causal_two_phase_launch(
        contextRef,
        elementDType,
        "iv_estimate",
        partialCount,
        partialArgs,
        sizeof(partialArgs),
        finalizeArgs,
        sizeof(finalizeArgs),
        completionToken,
        status
    );
}

File: device/metal/causal/native/intervention.m
#include "intervention.h"
#include "causal.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

static const NSUInteger metalCausalScalarThreadCount = 256;

int metal_dispatch_do_intervene(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef adjacencyRef,
    MetalBufferRef intervenedRef,
    MetalBufferRef outRef,
    uint32_t nodeCount,
    uint32_t intervenedCount,
    uint64_t completionToken,
    MetalStatus* status
) {
    if (adjacencyRef == NULL || intervenedRef == NULL || outRef == NULL) {
        metal_causal_status_set(status, -2, "nil Metal buffer");
        return -2;
    }

    return metal_causal_named_dispatch(
        contextRef, elementDType, "do_intervene", (NSUInteger)nodeCount * nodeCount,
        completionToken, status,
        ^(id<MTLComputeCommandEncoder> encoder) {
            [encoder setBuffer:(__bridge id<MTLBuffer>)adjacencyRef offset:0 atIndex:0];
            [encoder setBuffer:(__bridge id<MTLBuffer>)intervenedRef offset:0 atIndex:1];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:2];
            [encoder setBytes:&nodeCount length:sizeof(nodeCount) atIndex:3];
            [encoder setBytes:&intervenedCount length:sizeof(intervenedCount) atIndex:4];
        }
    );
}

int metal_dispatch_cate(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef treatedRef,
    MetalBufferRef controlRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    MetalStatus* status
) {
    if (treatedRef == NULL || controlRef == NULL || outRef == NULL) {
        metal_causal_status_set(status, -2, "nil Metal buffer");
        return -2;
    }

    return metal_causal_named_dispatch(
        contextRef, elementDType, "cate", (NSUInteger)count, completionToken, status,
        ^(id<MTLComputeCommandEncoder> encoder) {
            [encoder setBuffer:(__bridge id<MTLBuffer>)treatedRef offset:0 atIndex:0];
            [encoder setBuffer:(__bridge id<MTLBuffer>)controlRef offset:0 atIndex:1];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:2];
            [encoder setBytes:&count length:sizeof(count) atIndex:3];
        }
    );
}

int metal_dispatch_counterfactual(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef observedYRef,
    MetalBufferRef observedXRef,
    MetalBufferRef counterfactualXRef,
    MetalBufferRef slopeRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    MetalStatus* status
) {
    if (observedYRef == NULL || observedXRef == NULL || counterfactualXRef == NULL ||
        slopeRef == NULL || outRef == NULL) {
        metal_causal_status_set(status, -2, "nil Metal buffer");
        return -2;
    }

    return metal_causal_named_dispatch(
        contextRef, elementDType, "counterfactual", (NSUInteger)count, completionToken, status,
        ^(id<MTLComputeCommandEncoder> encoder) {
            [encoder setBuffer:(__bridge id<MTLBuffer>)observedYRef offset:0 atIndex:0];
            [encoder setBuffer:(__bridge id<MTLBuffer>)observedXRef offset:0 atIndex:1];
            [encoder setBuffer:(__bridge id<MTLBuffer>)counterfactualXRef offset:0 atIndex:2];
            [encoder setBuffer:(__bridge id<MTLBuffer>)slopeRef offset:0 atIndex:3];
            [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:4];
            [encoder setBytes:&count length:sizeof(count) atIndex:5];
        }
    );
}

static int metal_causal_two_phase_names(
    char* partialName,
    char* finalizeName,
    size_t nameBytes,
    const char* operationName,
    int elementDType,
    MetalStatus* status
) {
    char baseName[128];
    int nameCode = metal_causal_kernel_name(
        baseName, sizeof(baseName), operationName, elementDType, status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    int partialWritten = snprintf(partialName, nameBytes, "%s_partial", baseName);
    int finalizeWritten = snprintf(finalizeName, nameBytes, "%s_finalize", baseName);

    if (partialWritten <= 0 || finalizeWritten <= 0 ||
        (size_t)partialWritten >= nameBytes || (size_t)finalizeWritten >= nameBytes) {
        metal_causal_status_set(status, -6, "Metal causal scalar kernel name overflow");
        return -6;
    }

    return 0;
}

static int metal_causal_two_phase_pipelines(
    MetalContext* context,
    const char* partialName,
    const char* finalizeName,
    MetalStatus* status,
    id<MTLComputePipelineState>* partialPipeline,
    id<MTLComputePipelineState>* finalizePipeline
) {
    int partialCode = metal_causal_pipeline(context, partialName, status, partialPipeline);

    if (partialCode != 0) {
        return partialCode;
    }

    return metal_causal_pipeline(context, finalizeName, status, finalizePipeline);
}

static int metal_causal_encode_iv_partial(
    id<MTLCommandBuffer> commandBuffer,
    id<MTLComputePipelineState> pipeline,
    MetalBufferRef instrumentRef,
    MetalBufferRef treatmentRef,
    MetalBufferRef outcomeRef,
    MetalBufferRef scratchRef,
    uint32_t count,
    uint32_t partialCount,
    MetalStatus* status
) {
    id<MTLComputeCommandEncoder> encoder = nil;
    int code = metal_causal_encoder(commandBuffer, pipeline, status, &encoder);

    if (code != 0) {
        return code;
    }

    [encoder setBuffer:(__bridge id<MTLBuffer>)instrumentRef offset:0 atIndex:0];
    [encoder setBuffer:(__bridge id<MTLBuffer>)treatmentRef offset:0 atIndex:1];
    [encoder setBuffer:(__bridge id<MTLBuffer>)outcomeRef offset:0 atIndex:2];
    [encoder setBuffer:(__bridge id<MTLBuffer>)scratchRef offset:0 atIndex:3];
    [encoder setBytes:&count length:sizeof(count) atIndex:4];
    [encoder
        dispatchThreadgroups:MTLSizeMake((NSUInteger)partialCount, 1, 1)
        threadsPerThreadgroup:MTLSizeMake(metalCausalScalarThreadCount, 1, 1)
    ];
    [encoder endEncoding];

    return 0;
}

static int metal_causal_encode_iv_finalize(
    id<MTLCommandBuffer> commandBuffer,
    id<MTLComputePipelineState> pipeline,
    MetalBufferRef scratchRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint32_t partialCount,
    MetalStatus* status
) {
    id<MTLComputeCommandEncoder> encoder = nil;
    int code = metal_causal_encoder(commandBuffer, pipeline, status, &encoder);

    if (code != 0) {
        return code;
    }

    [encoder setBuffer:(__bridge id<MTLBuffer>)scratchRef offset:0 atIndex:0];
    [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:1];
    [encoder setBytes:&count length:sizeof(count) atIndex:2];
    [encoder setBytes:&partialCount length:sizeof(partialCount) atIndex:3];
    [encoder
        dispatchThreadgroups:MTLSizeMake(1, 1, 1)
        threadsPerThreadgroup:MTLSizeMake(metalCausalScalarThreadCount, 1, 1)
    ];
    [encoder endEncoding];

    return 0;
}

int metal_dispatch_iv_estimate(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef instrumentRef,
    MetalBufferRef treatmentRef,
    MetalBufferRef outcomeRef,
    MetalBufferRef scratchRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint32_t partialCount,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        if (instrumentRef == NULL || treatmentRef == NULL || outcomeRef == NULL ||
            scratchRef == NULL || outRef == NULL) {
            metal_causal_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        char partialName[128];
        char finalizeName[128];
        int nameCode = metal_causal_two_phase_names(
            partialName, finalizeName, sizeof(partialName), "iv_estimate", elementDType, status
        );

        if (nameCode != 0) {
            return nameCode;
        }

        MetalContext* context = NULL;
        id<MTLCommandBuffer> commandBuffer = nil;
        int prepareCode = metal_causal_prepare(contextRef, status, &context, &commandBuffer);

        if (prepareCode != 0) {
            return prepareCode;
        }

        id<MTLComputePipelineState> partialPipeline = nil;
        id<MTLComputePipelineState> finalizePipeline = nil;
        int pipelineCode = metal_causal_two_phase_pipelines(
            context, partialName, finalizeName, status, &partialPipeline, &finalizePipeline
        );

        if (pipelineCode != 0) {
            return pipelineCode;
        }

        int partialCode = metal_causal_encode_iv_partial(
            commandBuffer, partialPipeline, instrumentRef, treatmentRef, outcomeRef,
            scratchRef, count, partialCount, status
        );

        if (partialCode != 0) {
            return partialCode;
        }

        int finalizeCode = metal_causal_encode_iv_finalize(
            commandBuffer, finalizePipeline, scratchRef, outRef, count, partialCount, status
        );

        if (finalizeCode != 0) {
            return finalizeCode;
        }

        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        [commandBuffer commit];
        return 0;
    }
}
